# Day 10: PCA: Unit testing

### <p style='text-align: right;'> &#9989; ARNAB MUSTAFI ARKA




---

# __Unit Testing: pre-class__

# ___Learning objectives___

At the end of the exercise, you should be able to:
- Explain why unit testing is needed.
- Explain what the unittest module is for.
- Implement a simple unit test.

# __Outline__ 


1. [Introduction to Unit Testing](#intro)
1. [Unit test example](#example)
1. [Assignment wrap-up](#wrapup)


---

<a id="intro"></a>

# ___Introduction to Unit Testing___

## What a unit test is:

Unit tests are small tests of individual parts of your code. 
- Absolutely necessary to grow a project past a few developers
- Ideally, they run after every major/minor change and provide a reality check that nothing is broken.
- Good unit tests are hard to do and can take practice and time (which is not often where you want to spend your time).  
- That being said, if you know the basic format/syntax of some of the most common testing programs you can format your code in advance to be ready for unit testing.  

🗒️ **Task:** In [the wikipedia page for unit testing](https://en.wikipedia.org/wiki/Unit_testing), the advantages and disadvantages for unit tests are discussed. In the space below, summarize the pros and cons of unit tests.

✏️ **Answer:** Unit tests are helpful because they find bugs early, make code easier to change, improve reliability, and support teamwork. However, they also take time to write, are hard to cover everything, often need updates, and are less useful if written poorly.

## The `unittest` module

🗒️ **Task:** We will start with the most basic unit test program built with python; ```unittest```. Make sure this is installed in your `cmse802` conda environment and run the following: 

In [16]:
# Activate your cmse802 environment

import unittest

help(unittest.TestCase.assertEqual)

Help on function assertEqual in module unittest.case:

assertEqual(self, first, second, msg=None)
    Fail if the two objects are unequal as determined by the '=='
    operator.



🗒️ **Question:** In the above line, the help information for `assertEqual` from `unittest.TestCase` is printed out. Answer the following questions:

- How is `assertEqual` different from `assertAlmostEqual`?
- What does `assertRaises` do?

Provide the code to get help from these two functions, then put your answers as comments in the code cell.

✏️ **Answer:**

In [17]:
### ANSWER
import unittest

# Get help for the two functions
help(unittest.TestCase.assertAlmostEqual)
help(unittest.TestCase.assertRaises)

### ANSWER

# assertEqual checks if two values are exactly equal (using ==).
# assertAlmostEqual checks if two values are approximately equal, useful for floating-point comparisons.

# assertRaises checks if a specific exception is raised when running code.


Help on function assertAlmostEqual in module unittest.case:

assertAlmostEqual(self, first, second, places=None, msg=None, delta=None)
    Fail if the two objects are unequal as determined by their
    difference rounded to the given number of decimal places
    (default 7) and comparing to zero, or by comparing that the
    difference between the two objects is more than the given
    delta.
    
    Note that decimal places (from zero) are usually not the same
    as significant digits (measured from the most significant digit).
    
    If the two objects compare equal then they will automatically
    compare almost equal.

Help on function assertRaises in module unittest.case:

assertRaises(self, expected_exception, *args, **kwargs)
    Fail unless an exception of class expected_exception is raised
    by the callable when invoked with specified positional and
    keyword arguments. If a different type of exception is
    raised, it will not be caught, and the test case will be
   

---
<a id="example"></a>
# ___Unit test example___

The code you develop below __will be used for an in-class exercise__. Please make sure you get them done!

## How to create unit test with `unittest`?

🗒️ **Task:** Watch [this 8 min video](https://www.youtube.com/watch?v=1Lfv5tUGsn8) titled _Unit Tests in Python_.

## Set up the function we want to test

🗒️ **Task:** Follow the examples provided in the video:

- Create a file named `circles.py` with a barebone `circle_area` function.  Store it in the same directory as this notebook.

The following code imports your `circle_area` function and calls it, passing $1$ as the argument. 

In [18]:
%%writefile circles.py
import math

def circle_area(r):
    return math.pi * (r ** 2)


Overwriting circles.py


In [19]:
%%writefile test_circles.py
import unittest, math
from circles import circle_area

class TestCircleArea(unittest.TestCase):
    def test_area(self):
        self.assertAlmostEqual(circle_area(1), math.pi, places=12)

if __name__ == "__main__":
    unittest.main()


Overwriting test_circles.py


In [20]:
from circles import circle_area

print(circle_area(1))

3.141592653589793


🗒️ **Task:** Now call `circle_area` by passing $-1$ as the argument. Explain if this result makes sense.

In [21]:
### ANSWER
from circles import circle_area

print(circle_area(-1))   


3.141592653589793


Explanation: The result is not meaningful because a circle cannot have a negative radius. The current barebone function simply squares the input and returns a positive area. A better version should raise an error when the radius is negative.

## Create `test_area` function

In unit tests, one of the first things we should do is to ensure the function you implmented gets the right answer.
- For example, for `circle_area`, you'd expect when the radius is 1, the answer should be ~3.14.

🗒️ **Task:** Follow the examples provided in the video:

- Create `test_circles.py` that use `unittest` and has just the `test_area` function.
- Run unit test in your terminal with the following command. What is the output from the test? Paste it below.

```bash
python -m unittest test_circles
```

In [ ]:
### ANSWER
import unittest, math
from circles import circle_area

class TestCircleArea(unittest.TestCase):
    def test_area(self):
        # Test when radius = 1
        self.assertAlmostEqual(circle_area(1), math.pi, places=12)

if __name__ == "__main__":
    unittest.main()


## Create `test_values` function

Once the test for evaluating the expected values can be generated, we can next focus on testing situations when the function should not work.

🗒️ **Task:** Follow the examples provided in the video:

- Revise your `test_circles.py` to include a `test_values` function that raises errors when the passed value is negative.
- Run a unit test. What is the output from the test? Paste it below.

✏️ **Answer:** test_area (test_circles.TestCircleArea.test_area) ... ok
test_values (test_circles.TestCircleArea.test_values) ... ok

Ran 2 tests in 0.001s

OK


In [ ]:
import unittest, math
from circles import circle_area

class TestCircleArea(unittest.TestCase):
    def test_area(self):
        # Test when radius = 1
        self.assertAlmostEqual(circle_area(1), math.pi, places=12)

    def test_values(self):
        # Test negative radius should raise an error
        with self.assertRaises(ValueError):
            circle_area(-1)

if __name__ == "__main__":
    unittest.main()


## Modify your `circles.py`

🗒️ **Task:** Through unit test, you know that the `circle_area` function cannot deal with situations when:

- A negative real number is provided,
- Or a non-real number is provided.

Revise your `circles.py` that will deal with the above two situations:
- Note that the tutorial video provides examples on how to do this.
- Put your revised `circles.py` in the space below.

In [ ]:
### ANSWER
import math

def circle_area(r):
    # Check if r is a number (int or float)
    if not isinstance(r, (int, float)):
        raise TypeError("The radius must be a real number (int or float).")
    
    # Check if r is negative
    if r < 0:
        raise ValueError("The radius cannot be negative.")
    
    return math.pi * (r ** 2)


## Ensure that it passes your tests

Re-run the `unittest` command to ensure that all tests now pass.  How satisfying!

---
## Assignment wrap-up
Please fill out the form that appears when you run the code below.  **You must completely fill this out in order to receive credit for the assignment!** If running the cell doesn't work in VS Code copy the link `src` and paste in the browser. Make sure to sign in with your MSU email. 


In [ ]:
from IPython.display import HTML
HTML(
'''
<iframe 
    src="https://forms.office.com/r/A9kij6yedJ" 
    width="800px" 
    height="600px" 
    frameborder="0" 
    marginheight="0" 
    marginwidth="0">
    Loading...
</iframe>
'''
)


## Congratulations, you're done with your pre-class assignment!

Now, you just need to submit this assignment by uploading it to the course <a href="https://d2l.msu.edu/">Desire2Learn</a> web page for the appropriate pre-class submission folder. (Don't forget to add your name in the first cell).
